### PR TITLE
feat(mongodb): collect mongodb4 metrics

### LINKED ISSUE #1157: Metrics instrumentaion MongoDB
Hey, we want to start to implement metrics collection in the instrumentations like other SDKs already did.
I'll start with MongoDB according to:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/database-metrics.md

If you find it right to start implementing metrics collection, please assign this issue to me :)

### PR DESCRIPTION
**Pull request by osherv**
## Which problem is this PR solving?

-

## Short description of the changes
Fixes #1157 

Added metrics collection for MongoDB >= 4, according to: 
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/database-metrics.md
This is the first of many contributions planned under metrics collections at the instrumentation level, so I'll be really happy to get as much as comments as we can on this :)
Thanks!

## Checklist

- [ ] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

**Comment by vmarchaud** [comment]
Do you want to wait for https://github.com/open-telemetry/opentelemetry-js/pull/3267 to merge this ?

**Comment by vmarchaud** [comment]
Converting to draft in the mean time :)

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (haddasbronfman) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:867]: Comment to myself and others - I think until this pr: https://github.com/open-telemetry/opentelemetry-specification/pull/3050 is not approved, we can't approve the pool name issue. (This pr wants to define the pool.name to be the db.connection_string in case the DB doesn't supply pool name).

**Comment by haddasbronfman** [comment]
@open-telemetry/javascript-approvers Can someone please review?

File: plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts:91]: Can this and `V4Connect` be an internal type instead? :thinking:

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:65]: Not sure if the one in the semconv is a typo or this one here (see https://github.com/open-telemetry/opentelemetry-specification/blob/c210cf5884cf2c06dc1d1a6a34a4ea76c542841f/specification/metrics/semantic_conventions/database-metrics.md?plain=1#L34) :slightly_smiling_face: 
```suggestion
        unit: '{connection}',
```

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:64]: We've recently had an issue in the core repo (https://github.com/open-telemetry/opentelemetry-js/pull/3710) where the prometheus collector exporter would complain when the description does not exactly match the semconv and other language implementations are.

It's a bit unfortunate that the description has to be followed exactly even though it is not specified as identifying, but I think this could save us some maintenance later :slightly_smiling_face: 

```suggestion
          'The number of connections that are currently in state described by the state attribute.',
```

**Review by pichlermarc — COMMENTED** [review]
Thanks for adding this, looking good :slightly_smiling_face: 
Just a few question/suggestions around types.

Looks like the failing tests may be unrelated, it may be happening on main too. :thinking:

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (haddasbronfman) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:65]: I looked [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-units) and it says: "All non-units that use curly braces to annotate a quantity need to match the grammatical number of the quantity it represent. For example if measuring the number of individual requests to a process the unit would be {request}, not {requests}.".
If I understand correctly we should use {connection} and not {connections}. what do you think?

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (haddasbronfman) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:64]: two questions about this:
1. When you say 'exactly' it means that 'state' should appear with single quotes? I mean `'state'` and not `state` as in semconv. 
2. This is relevant also for mysql, right? see [here](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1220#discussion_r1046851794), where we had a long discussion about this :)

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (pichlermarc) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:65]: Ah yes, that's also what I understand when reading the spec there. I agree, I think we should switch to `{connection}`. :slightly_smiling_face:

Comment (haddasbronfman) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts]: I should change it in mysql too https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1220

**Comment by haddasbronfman** [comment]
@legendecas @dyladan  Hi. I'm looking for reviewers who are familiar with metrics. Can you please review this PR?

File: plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts:122]: Could it be that this test is not being run? :thinking: 
The description has changed, so this should fail but [the job](https://github.com/open-telemetry/opentelemetry-js-contrib/actions/runs/4853420360/jobs/8649566069) somehow passes. :thinking: 
[TAV fails, though.](https://github.com/open-telemetry/opentelemetry-js-contrib/actions/runs/4853420381/jobs/8649578432?pr=1170)

Comment (haddasbronfman) [reply plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts:122]: you're right. look at the [tav.yml](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-mongodb/.tav.yml):
```
mongodb:
  jobs:
    - versions: ">=3.3 <4"
      commands: npm run test
    - versions: ">=4 <5"
      commands: npm run test-new-versions
``` 
 version 4 tests will run by the command 'test-new-versions' and version 3 tests will run by the command 'npm run test'. This is why you don't see the v4 test in the unit-test job. 

I'll fix the description issue so TAV will pass. 

But there is something still bothering me: in unit-test job it seems that v3 test didn't run at all:
image
I'll take a look at this.

File: plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts
Comment (legendecas) [inline plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts:37]: Nit: IIUC, a pulling metric reader might be better suited for testing usage since it can pull metrics proactively instead of awaiting timeouts.

**Review by legendecas — APPROVED** [review]
I'm not familiar with mongodb. The metrics instrumentation looks good to me.

File: plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts
Comment (pichlermarc) [inline plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts:124]: looks like tav is failing here now :thinking: 
```suggestion
    assert.strictEqual(metrics[0].descriptor.unit, '{connection}');
```

Comment (haddasbronfman) [reply plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts:124]: at least we have TAV. it helps us don't miss anything :)

File: plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts
Comment (pichlermarc) [reply plugins/node/opentelemetry-instrumentation-mongodb/test/mongodb-v4.metrics.test.ts]: Marking this as resolved as the tests are also not running on main. Thank you for creating the issue (#1487) :slightly_smiling_face:

**Comment by pichlermarc** [comment]
Looks like TAV is failing on line 137 in the v4 metrics tests, the idle state is now exporting `2`. :thinking:

**Pull request merged**
